larray/core/session.py

```diff
@@ -1475,6 +1478,164 @@ def display(k, v, is_metadata=False):
+    ...     # Their values will be defined at runtime but must match the specified type.
+    ...     variant_name = str
+    ...     first_proj_year = int
+    ...     obs_years = Axis
+    ...     proj_years = Axis
+    ...     population_obs = Array
+    ...     population_proj = Array
+    ...     # declare arrays with fixed axes
+    ...     mortality_rate = ArrayDef((AGE, GENDER))
+
+    >>> def run_model(variant_name, first_proj_year):
+    ...     # create an instance of the ModelVariables class
+    ...     m = ModelVariables()
+    ...     # ==== setup variables ====
+    ...     # set scalars
+    ...     m.variant_name = variant_name
+    ...     m.first_proj_year = first_proj_year
+    ...     # set axes
+    ...     m.obs_years = Axis(f'time={m.FIRST_OBS_YEAR}..{m.first_proj_year-1}')
+    ...     m.proj_years = Axis(f'time={m.first_proj_year}..{m.LAST_PROJ_YEAR}')
+    ...     # set arrays
+    ...     m.population_obs = zeros((m.AGE, m.GENDER, m.obs_years))
+    ...     m.population_proj = zeros((m.AGE, m.GENDER, m.proj_years))
+    ...     m.mortality_rate = zeros((m.AGE, m.GENDER))
+    ...     # ==== model ====
+    ...     # some code here
+    ...     # ...
+    ...     # ==== output ====
+    ...     # save all variables in an HDF5 file
+    ...     m.save('{variant_name}.h5', display=True)
+
+    Content of file 'main.py'
+
+    >>> run_model('projection_2020_2070', first_proj_year=2020)
+    dumping FIRST_OBS_YEAR ... done
+    dumping LAST_PROJ_YEAR ... done
+    dumping AGE ... done
+    dumping GENDER ... done
+    dumping G_CHILDREN ... done
+    dumping G_ADULTS ... done
+    dumping variant_name ... done
+    dumping first_proj_year ... done
+    dumping obs_years ... done
+    dumping proj_years ... done
+    dumping population_obs ... done
+    dumping population_proj ... done
+    dumping mortality_rate ... done
+    """
+    def __init__(self, *args, **kwargs):
+        _cls_attrs = {key: value for key, value in vars(self.__class__).items() if not key.startswith('_')}
+        object.__setattr__(self, '_cls_attrs', _cls_attrs)
+
+        constant_cls_attrs = {key: value for key, value in _cls_attrs.items()
+                              if not isinstance(value, (type, ArrayDef))}
+
+        if len(args) == 1:
+            # Note: using the two lines below we may loose the order in which variables have been declared
+            # but can we avoid it?
+            Session.__init__(self, *args, **kwargs)
+            self.add(**constant_cls_attrs)
+        else:
+            # Note: same as above
+            kwargs.update(constant_cls_attrs)
+            Session.__init__(self, *args, **kwargs)
+
+    def load(self, fname, names=None, engine='auto', display=False, **kwargs):
+        super().load(fname, names, engine, display, **kwargs)
+        for key in set(self._cls_attrs.keys()) - set(self.keys()):
+            warnings.warn(f"The variable {key} is declared in the {self.__class__.__name__} "
```

Comment:
I am **unsure** this warning should exist. Creating a session out of several files seems like something we should allow without a warning. Maybe the warning should be there only if names is None???

If key was in names and it's not in the file, this should be an error (not a warning) but I suppose super().load would already catch that.

Reply:
This is not innocent though. If we allow this, we must handle the case where we request an attribute which has been declared but not loaded (yet). I am leaning towards that being a good idea (because we will need something like that for LazySession anyway).

Reply:
But, to keep things simple, we might want to start without that feature and implement it later.